### PR TITLE
adding support for Cyclone V SoC DK ethernet

### PR DESCRIPTION
**Pull request by mevalver**
Adding support for Cyclone V SoC ethernet device

Signed-off-by: Esteban Valverde <esteban.valverde.vega@intel.com>

### CODE REVIEW

File: boards/arm/cyclonev_socdk/Kconfig.defconfig [hunk 1]
```diff
@@ -14,6 +14,11 @@ config I2C_DW_CLOCK_SPEED
 	default 200
 endif # I2C_DW
 
+if EEPROM
```
Comment (teburd) [inline boards/arm/cyclonev_socdk/Kconfig.defconfig:17]: This seems unrelated to the ethernet driver to me

Comment (mevalver) [reply boards/arm/cyclonev_socdk/Kconfig.defconfig:17]: Resolved

**Comment by mevalver** [comment]
@teburd I did a small modification to remove the EEPROM definition, now some automatic checks are not passing, but I don't understand the reason.

Is there anything else to be reviewed in this PR?

**Comment by teburd** [comment]
I rebased in hope that it solves the problem, but this means every commit now has my name on it (probably less than ideal). If it solves the CI issue please rebase on the latest main and force push again, once CI is green I'll +1

**Comment by mevalver** [comment]
@teburd seems like the test-9 keeps failing... BTW, I don't mind your name in the comments

**Comment by teburd** [comment]
This seems to be a known issue with logging, noted and shouldn't prevent a merge and approval here.

**Review by teburd — APPROVED** [review]

**Comment by mevalver** [comment]
@pfalcon, @tbursztyka, @MaureenHelm, @mbolivar I'd appreciate if one of you find a little time to give me a thumbs up in this PR or request any change you seem fit.

**Comment by mevalver** [comment]
@pfalcon, @tbursztyka, @MaureenHelm, @mbolivar I'd appreciate if one of you find a little time to give me a thumbs up in this PR or request any change you seem fit. (?)

**Review by teburd — DISMISSED** [review]

**Comment by teburd** [comment]
Rebased as the CI checks should be passing now

**Comment by mevalver** [comment]
CIs are passing now

**Review by teburd — DISMISSED** [review]

**Comment by mevalver** [comment]
@pfalcon, @tbursztyka, @MaureenHelm, @mbolivar seems like the checks are passing now, and I#ll require some helop from any of you to make this go. Thanks!

File: drivers/ethernet/eth_cyclonev.c [hunk 2]
```diff
@@ -0,0 +1,1174 @@
...
+
+/*	 4b. Disable the Ethernet Controller FPGA interfaces by clearing the emac_* bit in the
+ * module register of the System Manager (FPGA Interface group).
+ */
+
+	sys_clear_bits(SYSMGR_FPGAINTF_INDIV_ADDR, Sysmgr_Fpgaintf_En_3_Emac_Set_Msk[instance]);
+
+/*	 7. After confirming the settings are valid, software can clear the emac bit in the
+ * permodrst register of the Reset Manager to bring the EMAC out of reset.
+ */
```
Comment (MaureenHelm) [inline drivers/ethernet/eth_cyclonev.c:90]: Please indent the comment blocks

File: drivers/ethernet/eth_cyclonev.c [hunk 3]
```diff
@@ -0,0 +1,1174 @@
...
+
+	ret = eth_cyclonev_software_reset(p->instance);
+	if (ret != 0) {
+		LOG_ERR("eth_cyclonev_software_reset failure!\n");
+		return ret;
+	}
+
+/* 4. Program the following fields to initialize the Bus Mode Register by setting
+ * values in DMA Register 0 (Bus Mode Register):
+ */
```
Comment (MaureenHelm) [inline drivers/ethernet/eth_cyclonev.c:921]: Please indent the comment blocks

File: drivers/ethernet/eth_cyclonev.c [hunk 4]
```diff
@@ -0,0 +1,1174 @@
...
+const struct ethernet_api eth_cyclonev_api = {
+	.iface_api.init	= eth_cyclonev_iface_init,
+	.get_capabilities	= eth_cyclonev_caps,
+	.send		= eth_cyclonev_send,
+	.start		= eth_cyclonev_start,
+	.stop		= eth_cyclonev_stop,
+	.set_config	= eth_cyclonev_set_config
+};
+
+ETH_NET_DEVICE_DT_INST_DEFINE(0,
```
Comment (MaureenHelm) [inline drivers/ethernet/eth_cyclonev.c:1165]: Please use `DT_INST_FOREACH_STATUS_OKAY`

File: drivers/ethernet/phy_cyclonev.c [hunk 5]
```diff
@@ -0,0 +1,367 @@
...
+#define PHY_LINKED_STATUS		(1 << 2)
+#define PHY_AUTONEGO_COMPLETE		(1 << 5)
+
+/* Auto-Negotiation Advertisement */
+#define PHY_AUTON			(4)
+#define PHYANA_10BASET			(1 << 5)
+#define PHYANA_10BASETFD		(1 << 6)
+#define PHYANA_100BASETX		(1 << 7)
+#define PHYANA_100BASETXFD		(1 << 8)
+#define PHYASYMETRIC_PAUSE		(1 << 11)
```
Comment (MaureenHelm) [inline drivers/ethernet/phy_cyclonev.c:59]: Use `BIT()`

File: boards/arm/cyclonev_socdk/Kconfig.defconfig [hunk 6]
```diff
@@ -28,4 +28,14 @@ config SPI_DW_FIFO_DEPTH
...
 endif # SPI_DW
 
+if NETWORKING
+
+config NET_L2_ETHERNET
+	default y
+config TEST_RANDOM_GENERATOR
+	default y
+config ETH_CYCLONEV
+	default y
```
Comment (MaureenHelm) [inline boards/arm/cyclonev_socdk/Kconfig.defconfig:38]: This won't be needed after you make use of `DT_HAS_SNPS_ETHERNET_CYCLONEV_ENABLED`

**Review by MaureenHelm — CHANGES_REQUESTED** [review]

**Comment by teburd** [comment]
There's some weird indentation going on, I'd recommend running clang-format on the files in this PR.

**Comment by mevalver** [comment]
@teburd clang-format destroyed the format of my file.... I need time to fix the issues again :(

**Comment by mevalver** [comment]
@teburd. @MaureenHelm all the CIs are passing now, and all the different changes you suggested are now in place!
Let me know if anything else is required.

**Review by teburd — DISMISSED** [review]

File: dts/bindings/ethernet/snps,ethernet-cyclonev.yaml [hunk 7]
```diff
@@ -0,0 +1,17 @@
+# SPDX-License-Identifier: Apache-2.0
+# Copyright (C) 2022, Intel Corporation
+# Description:
+# Binding file for Ethernet driver CycloneV
+# heavily modified for Zephyr
```
Comment (mbolivar-nordic) [inline dts/bindings/ethernet/snps,ethernet-cyclonev.yaml:5]: What is this for?

Also, please improve the commit message. `dts: arm:` should be something more descriptive, such as `dts: intel_socfgpa:` or so.

**Review by mbolivar-nordic — CHANGES_REQUESTED** [review]
Couple of small requests for the binding. I did not review the rest.

BTW, zephyr pings to my personal `mbolivar` github handle are currently ignored, since I switched to using a work-specific account a couple of years ago.

**Review by teburd — DISMISSED** [review]

File: drivers/ethernet/eth_cyclonev.c [hunk 8]
```diff
@@ -0,0 +1,1166 @@
...
+ * @brief Inits device structure variables
+ * Saves base address into variable, detects device instance number
+ *
+ * @param p Pointer to device structure
+ * @retval 0 if successful, -1 otherwise
+ */
+
+int eth_cyclonev_bus_init(struct eth_cyclonev_priv *p)
+{
+	p->base_addr = DT_INST_REG_ADDR(0);
```
Comment (MaureenHelm) [inline drivers/ethernet/eth_cyclonev.c:786]: Please don't hard code instance numbers in interface functions

File: drivers/ethernet/eth_cyclonev_priv.h [hunk 9]
```diff
@@ -0,0 +1,509 @@
...
+	uint8_t tx_buf[ETH_BUFFER_SIZE * NB_TX_DESCS]; /* Transmit Buffer */
+
+	struct k_sem free_tx_descs;
+	uint8_t running;     /* Running state flag */
+	uint8_t initialised; /* Initialised state flag */
+};
+
+/* Emac Base Addresses */
+#define EMAC0_ADDR 0xFF700000
+#define EMAC1_ADDR 0xFF702000
```
Comment (MaureenHelm) [inline drivers/ethernet/eth_cyclonev_priv.h:56]: Base addresses here and elsewhere in this file need to come from devicetree

**Review by MaureenHelm — CHANGES_REQUESTED** [review]

**Review by MaureenHelm — DISMISSED** [review]

**Review by teburd — DISMISSED** [review]

**Comment by teburd** [comment]
@mbolivar-nordic seems your comments have been addressed, please take a look when you can

File: boards/arm/cyclonev_socdk/Kconfig.defconfig [hunk 10]
```diff
@@ -28,4 +28,13 @@ config SPI_DW_FIFO_DEPTH
 	default 256
 endif # SPI_DW
 
+if NETWORKING
```
Comment (mbolivar-nordic) [inline boards/arm/cyclonev_socdk/Kconfig.defconfig:31]: The commit message is unrelated to the change; please fix whatever is wrong here

**Review by mbolivar-nordic — CHANGES_REQUESTED** [review]

**Comment by mevalver** [comment]
@mbolivar-nordic @MaureenHelm @teburd again, changes were applied as suggested, please review/approve

**Review by teburd — DISMISSED** [review]

**Comment by mevalver** [comment]
@mbolivar-nordic could you please help me to get this going?

**Comment by mbolivar-nordic** [comment]
We are in feature freeze right now; what are you expecting? This won't be merged until the v3.3 merge window opens up anyway...

**Comment by teburd** [comment]
This needs a rebase now that 3.2 is out

**Review by teburd — APPROVED** [review]

**Review by MaureenHelm — APPROVED** [review]

**Comment by MaureenHelm** [comment]
@mbolivar-nordic please revisit

**Review by mbolivar-nordic — APPROVED** [review]

**Comment by mbolivar-nordic** [comment]
Thanks for the patch and sorry for the delay. Really appreciate your patience!
